### PR TITLE
encode opus with ffmpeg

### PR DESCRIPTION
**Pull request by davidcole1340**
woohoo no more dca required, we can just encode opus with ffmpeg. why didn't we do this in the beginning???? also fixed the choppy audio bug in the process

### CODE REVIEW

File: src/Discord/Exceptions/BufferTimedOutException.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+namespace Discord\Exceptions;
+
+use Exception;
+
+/**
+ * Thrown when reading from a buffer times out.
+ *
+ * @since 10.0.0
+ */
+class BufferTimedOutException extends Exception
```
Comment (SQKo) [inline src/Discord/Exceptions/BufferTimedOutException.php:21]: should extends RuntimException

File: src/Discord/Voice/OggPage.php [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+use Exception;
+use Discord\Helpers\Buffer;
+use Generator;
+use React\Promise\ExtendedPromiseInterface;
+
+/**
+ * Represents a page in an Ogg container.
+ *
+ * @see https://www.rfc-editor.org/rfc/rfc3533
```
Comment (SQKo) [inline src/Discord/Voice/OggPage.php:22]:
```suggestion
 * @link https://www.rfc-editor.org/rfc/rfc3533
```

File: src/Discord/Voice/OggPage.php [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+     * @param int    $version         The version number of the Ogg file format used in this stream.
+     * @param int    $headerType      Identifies the specific type of this page.
+     * @param int    $granulePosition Contains position information.
+     * @param int    $bitstreamSn     Contains the unique serial number by which the logical bitstream is identified.
+     * @param int    $pageSeq         Contains the sequence number of the page so the decoder can identify page loss.
+     * @param int    $checksum        Contains a 32 bit CRC checksum of the page (including header with zero CRC field and page content).
+     * @param int[]  $pageSegments    A list of page segment lengths which were contained within the page.
+     * @param string $segmentData     The data of all the page segments concatenated.
+     *
+     * @see https://www.rfc-editor.org/rfc/rfc3533 -- Section 6. The Ogg page format
```
Comment (SQKo) [inline src/Discord/Voice/OggPage.php:49]:
```suggestion
     * @link https://www.rfc-editor.org/rfc/rfc3533 Section 6. The Ogg page format
```

File: src/Discord/Voice/OggPage.php (hunk 2 above)
Comment (SQKo) [inline src/Discord/Voice/OggPage.php:22]:
```suggestion
 * @link https://www.rfc-editor.org/rfc/rfc3533
```

File: src/Discord/Voice/OggPage.php (hunk 3 above)
Comment (SQKo) [inline src/Discord/Voice/OggPage.php:49]:
```suggestion
     * @link https://www.rfc-editor.org/rfc/rfc3533#section-6 The Ogg page format
```

File: src/Discord/Voice/OggStream.php [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+use Exception;
+use React\Promise\ExtendedPromiseInterface;
+use React\Promise\Promise;
+
+use function React\Promise\resolve;
+
+/**
+ * Represents an Ogg Opus stream.
+ *
+ * @see https://www.rfc-editor.org/rfc/rfc7845.html
```
Comment (SQKo) [inline src/Discord/Voice/OggStream.php:25]:
```suggestion
 * @link https://www.rfc-editor.org/rfc/rfc7845
```

File: src/Discord/Voice/OpusHead.php [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+ */
+
+namespace Discord\Voice;
+
+use Exception;
+
+/**
+ * Represents the header attached to an Opus Ogg file.
+ *
+ * @see https://www.rfc-editor.org/rfc/rfc7845.html -- Section 5.1. Identification Header
```
Comment (SQKo) [inline src/Discord/Voice/OpusHead.php:19]:
```suggestion
 * @link https://www.rfc-editor.org/rfc/rfc7845#section-5.1 Identification Header
```

File: src/Discord/Voice/OpusTags.php [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+ */
+
+namespace Discord\Voice;
+
+use Exception;
+
+/**
+ * Represents Vorbis tags attached to an Opus Ogg file.
+ *
+ * @see https://www.rfc-editor.org/rfc/rfc7845.html -- Section 5.2. Comment Header
```
Comment (SQKo) [inline src/Discord/Voice/OpusTags.php:19]:
```suggestion
 * @link https://www.rfc-editor.org/rfc/rfc7845#section-5.2 Comment Header
```

**Review by SQKo — CHANGES_REQUESTED** [review]
use of @link instead of @see

File: src/Discord/Voice/OggPage.php [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+                );
+            });
+    }
+
+    /**
+     * Iterates through the packets contained within the stream, yielding an
+     * array containing binary data and whether the packet is complete, from
+     * a generator.
+     *
+     * @return Generator<int, (string|bool)[], mixed, void>
```
Comment (SQKo) [inline src/Discord/Voice/OggPage.php:118]:
```suggestion
     * @return Generator<int,(string|bool)[],mixed,void>
```

File: src/Discord/Voice/OggPage.php [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        return $buffer->read(4, timeout: $timeout)->then(function ($magic) use ($buffer, $timeout) {
+            if ($magic !== 'OggS') {
+                throw new Exception('Invalid Ogg page header, expected OggS got '.$magic);
+            }
+
+            return $buffer->read(23, timeout: $timeout);
+        })
+            // Reading header
+            ->then(function ($data) use ($buffer, &$header, $timeout) {
+                $header = unpack(Self::FORMAT, $data);
```
Comment (SQKo) [inline src/Discord/Voice/OggPage.php:87]:
```suggestion
                $header = unpack(self::FORMAT, $data);
```

File: src/Discord/Voice/OpusHead.php [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+     * @throws Exception If the binary data was missing the magic bytes.
+     */
+    public function __construct(string $data)
+    {
+        $magic = substr($data, 0, 8);
+        if ($magic != 'OpusHead') {
+            throw new Exception("Expected OpusHead, found $magic");
+        }
+
+        $head = unpack(Self::FORMAT, $data, 8);
```
Comment (SQKo) [inline src/Discord/Voice/OpusHead.php:122]:
```suggestion
        $head = unpack(self::FORMAT, $data, 8);
```

File: src/Discord/Voice/OpusHead.php [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+        $this->version = $head['version'];
+        $this->channelCount = $head['channel_count'];
+        $this->preSkip = $head['pre_skip'];
+        $this->sampleRate = $head['sample_rate'];
+        $this->outputGain = $head['output_gain'];
+        $this->channelMapFamily = $head['channel_map_family'];
+
+        if ($head['channel_map_family'] > 0) {
+            $stream_counts = unpack(Self::STREAM_COUNT_FORMAT, $data, 19);
```
Comment (SQKo) [inline src/Discord/Voice/OpusHead.php:131]:
```suggestion
            $stream_counts = unpack(self::STREAM_COUNT_FORMAT, $data, 19);
```

**Review by SQKo — CHANGES_REQUESTED** [review]
code editor friendly formatting

**Review by key2peace — DISMISSED** [review]

**Review by SQKo — APPROVED** [review]

**Pull request merged**